### PR TITLE
Minor changes to backend

### PR DESCRIPTION
**Pull request by hbs**

### CODE REVIEW

File: warp10/src/main/java/io/warp10/continuum/store/Store.java [hunk 1]
```diff
@@ -150,61 +188,31 @@ public class Store extends Thread {
...
 
-  private long throttlingDelay = 10000000L;
+  private final boolean FDBUseTenantPrefix;
 
   public Store(KeyStore keystore, final Properties properties, Integer nthr) throws IOException {
     this.keystore = keystore;
     this.properties = properties;
 
-    if ("true".equals(properties.getProperty("store.skip.write"))) {
+    if ("true".equals(properties.getProperty(Configuration.STORE_SKIP_WRITE))) {
```
Comment (Giwi) [inline warp10/src/main/java/io/warp10/continuum/store/Store.java:215]: Could be replaced by
`this.SKIP_WRITE = "true".equals(properties.getProperty(Configuration.STORE_SKIP_WRITE));`

File: warp10/src/main/java/io/warp10/continuum/store/Store.java [hunk 2]
```diff
@@ -691,63 +547,73 @@ private static class StoreConsumer implements Runnable {
...
     public void run() {
 
+      this.currentThread = Thread.currentThread();
       Thread.currentThread().setName("[Store Consumer - gen " + this.store.uuid + "/" + store.generation + "]");
 
+      // We need to call subscribe here otherwise the thread associated with 'consumer' will be the one which created the consumer
+      // and any call to poll or close will lead to an exception stating that the consumer is not MT safe
+      this.consumer.subscribe(this.topics);
+
       long count = 0L;
```
Comment (Giwi) [inline warp10/src/main/java/io/warp10/continuum/store/Store.java:611]: updated L930 but never used, could be deleted

File: warp10/src/main/java/io/warp10/fdb/FDBClear.java [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+import java.util.Arrays;
+
+import com.apple.foundationdb.Transaction;
+
+public class FDBClear implements FDBMutation {
+
+  private final byte[] tenant;
+  private final byte[] key;
+
+  public FDBClear(byte[] key) {
```
Comment (Giwi) [inline warp10/src/main/java/io/warp10/fdb/FDBClear.java:28]: Never used

File: warp10/src/main/java/io/warp10/fdb/FDBKVScanner.java [hunk 4]
```diff
@@ -0,0 +1,276 @@
...
+   */
+  private long offset = 1000000L;
+
+  private Long forceReadVersion = null;
+
+  private final FDBContext context;
+
+  private long count = 0L;
+
+  public FDBKVScanner(FDBContext fdbContext, Database db, FDBScan scan, StreamingMode mode) {
```
Comment (Giwi) [inline warp10/src/main/java/io/warp10/fdb/FDBKVScanner.java:75]: Never used

File: warp10/src/main/java/io/warp10/fdb/FDBMutation.java [hunk 5]
```diff
@@ -14,5 +14,18 @@
...
-namespace java io.warp10.continuum.thrift.service
+package io.warp10.fdb;
+
+import com.apple.foundationdb.Transaction;
+
+public interface FDBMutation {
+  /**
+   * Apply the mutation to a Transaction
+   */
+  public void apply(Transaction txn);
```
Comment (Giwi) [inline warp10/src/main/java/io/warp10/fdb/FDBMutation.java:25]: Modifier 'public' is redundant for interface members

File: warp10/src/main/java/io/warp10/fdb/FDBPool.java [hunk 6]
```diff
@@ -0,0 +1,66 @@
...
+ * So launching FoundationDB with the following environment variables set:
+ *
+ * export FDB_NETWORK_OPTION_EXTERNAL_CLIENT_LIBRARY=/path/to/libfdb_c.so
+ * export FDB_NETWORK_OPTION_CLIENT_THREADS_PER_VERSION=N
+ *
+ * will start N network threads, if the number of Database instances in the
+ * Warp 10 instance (sum of Databases created by all components) is <= N
+ * then each Database will have its own client.
+ *
+ * Experimenting is key as sizing depends n the type of load of each deployment.
```
Comment (Giwi) [inline warp10/src/main/java/io/warp10/fdb/FDBPool.java:38]: Typo "depends on"

File: warp10/src/main/java/io/warp10/fdb/FDBSet.java [hunk 7]
```diff
@@ -0,0 +1,57 @@
...
+public class FDBSet implements FDBMutation {
+
+  /**
+   * Tenant prefix
+   */
+  private final byte[] tenant;
+  private final byte[] key;
+  private final byte[] value;
+
+  public FDBSet(byte[] key, byte[] value) {
```
Comment (Giwi) [inline warp10/src/main/java/io/warp10/fdb/FDBSet.java:32]: Never used

File: warp10/src/main/java/io/warp10/fdb/FDBStoreClient.java [hunk 8]
```diff
@@ -0,0 +1,175 @@
...
+    // Some hints to choose the best plan:
+    // # of different classes
+    // # of different instances among each class
+    // # among those instances, common labels
+    //
+    // TODO(hbs): we might have to gather statistics to better determine the exec plan
+
+
+    //
+    // Sort metadatas so as to optimize the range scans
```
Comment (Giwi) [inline warp10/src/main/java/io/warp10/fdb/FDBStoreClient.java:130]: Typo: "Sort metadatas to optimize the range scans"

File: warp10/src/main/java/io/warp10/fdb/FDBUtils.java [hunk 9]
```diff
@@ -0,0 +1,260 @@
...
+    if (null != encoder) {
+      size += encoder.getPessimisticSize();
+      count += encoder.getCount();
+    }
+
+    size += count * (103 + 8 /* tenant */ + 8 /* class ID */ + 8 /* labels ID */ + 1 /* key prefix, we do not count the timestamp which is already accounted for */);
+
+    //
+    // If the estimated size is greater or equal to 4,500,000 (45 % of FDB max size), then we return true.
+    if (size >= 4500000) {
```
Comment (Giwi) [inline warp10/src/main/java/io/warp10/fdb/FDBUtils.java:254]: Could be replaced by: ` return size >= 4500000;`

File: warp10/src/main/java/io/warp10/hadoop/WritableUtils.java [hunk 10]
```diff
@@ -50,30 +51,30 @@
...
  */
 public class WritableUtils {
-  
+
   private static final boolean strictWritables;
   private static final boolean rawWritables;
-  
+
   static {
     if ("true".equals(WarpConfig.getProperty(Configuration.CONFIG_WARPSCRIPT_HADOOP_STRICTWRITABLES))) {
```
Comment (Giwi) [inline warp10/src/main/java/io/warp10/hadoop/WritableUtils.java:59]: Could be replaced by: `strictWritables = "true".equals(WarpConfig.getProperty(Configuration.CONFIG_WARPSCRIPT_HADOOP_STRICTWRITABLES));`

File: warp10/src/main/java/io/warp10/hadoop/WritableUtils.java [hunk 11]
```diff
@@ -50,30 +51,30 @@
...
       strictWritables = true;
     } else {
       strictWritables = false;
     }
-    
+
     if ("true".equals(WarpConfig.getProperty(Configuration.CONFIG_WARPSCRIPT_HADOOP_RAWWRITABLES))) {
```
Comment (Giwi) [inline warp10/src/main/java/io/warp10/hadoop/WritableUtils.java:65]: Could be replaced by: `
    rawWritables = "true".equals(WarpConfig.getProperty(Configuration.CONFIG_WARPSCRIPT_HADOOP_RAWWRITABLES));`

File: warp10/src/main/java/io/warp10/script/ScriptRunnerConsumerFactory.java [hunk 12]
```diff
@@ -16,94 +16,103 @@
...
-import io.warp10.crypto.CryptoUtils;
-import io.warp10.sensision.Sensision;
-
 import java.io.ByteArrayInputStream;
 import java.io.ByteArrayOutputStream;
 import java.io.OutputStream;
 import java.net.HttpURLConnection;
 import java.net.URL;
 import java.net.URLEncoder;
+import java.time.Duration;
```
Comment (Giwi) [inline warp10/src/main/java/io/warp10/script/ScriptRunnerConsumerFactory.java:25]: Unused imports

**Review by Giwi — CHANGES_REQUESTED** [review]

File: crypto/src/main/java/io/warp10/crypto/DummyKeyStore.java [hunk 13]
```diff
@@ -39,7 +39,7 @@ public DummyKeyStore() {
     // Initialize key store
     //
 
-    setKey(KeyStore.AES_HBASE_METADATA, new byte[32]);
+    setKey(KeyStore.AES_FDB_METADATA, new byte[32]);
```
[the remark this answers is not in the document]
Comment (hbs) [reply crypto/src/main/java/io/warp10/crypto/DummyKeyStore.java:42]: That could be a change we do later, but for now _FDB_ are to be considere related to the storage backend, vs _KAFKA_

File: crypto/src/main/java/io/warp10/crypto/KeyStore.java [hunk 14]
```diff
@@ -57,12 +57,12 @@ public interface KeyStore {
   /**
    * Name of key for wrapping metadata
    */
-  public static final String AES_HBASE_METADATA = "warp.aes.hbase.metadata";
+  public static final String AES_FDB_METADATA = "warp.aes.fdb.metadata";
```
[the remark this answers is not in the document]
Comment (hbs) [reply crypto/src/main/java/io/warp10/crypto/KeyStore.java:60]: ditto

File: warp10/src/main/java/io/warp10/continuum/KafkaWebCallService.java [hunk 15]
```diff
@@ -22,19 +22,19 @@
 import io.warp10.script.thrift.data.WebCallRequest;
 
 import java.util.Properties;
+import java.util.concurrent.Future;
```
[the remark this answers is not in the document]
Comment (hbs) [reply warp10/src/main/java/io/warp10/continuum/KafkaWebCallService.java:25]: removed in future PR

**Review by pi-r-p — APPROVED** [review]

File: warp10/src/main/java/io/warp10/script/ScriptRunnerConsumerFactory.java (hunk 12 above)
Comment (hbs) [reply warp10/src/main/java/io/warp10/script/ScriptRunnerConsumerFactory.java:25]: Leaving it for future use.

**Pull request merged**